### ISSUE TITLE
[Bug] Undo not available after aligning widgets [sf#68]

### ISSUE THREAD

**Issue by mantaraya36** [mantaraya36/CsoundQt#68]
**Reported by loucohen on 2010-08-25 15:38 UTC**
If I align a group of widgets incorrectly, for example selecting "align horizontally" when I should have selected "align vertically", there is no way to automatically recover from the error. The Undo command does not position the widgets to their original position. Worse, I cannot save the csd, change the position of widgets, then re-open or reload the saved csd, since apparently the widget information is written to the file even though I have not issued a SAVE.  Therefore there is no safety net available when I perform significant edits of widget positions.

**Comment by mantaraya36** [comment]
**Commented by mantaraya36 on 2010-08-25 17:25 UTC**
I am not seeing this behavior in svn, and I think there hasn't been a change in this area, but maybe it has been fixed. Can you provide exact steps to reproduce this problem?
Also, widgets are not saved automatically, except if you run the csd, and you have "save when run" enabled in the preferences. Also could you provide exact steps to reproduce this behavior?

**Comment by mantaraya36** [comment]
**Commented by loucohen on 2010-08-26 13:22 UTC**
Andres, after testing a little more systematically, I've found three things:
1\. After changing the configuration of widgets, it does NOT seem as if the widgets are being saved to the file. I was wrong about this, although it certainly seemed like that happened to me earlier. But I cannot reproduce it.

2\. If I move widgets around, and then Quit from QTCS, I do not get an offer to save the file. I think any time the file is changed in any way, the user should be offered the chance to save the file.

3\. Undo does not work the same way in all cases. It may be related to exactly what action was performed immediately before. In the steps below I found that UNDO sometimes did nothing \(depending on whether I used CMD-Z or EDIT/UNDO\), and sometimes UNDO jumped back two steps instead of one step. I hope the details below help.

-Lou


Open QTCS by double-clicking the application icon, confirm version \(0.6.0\).
In a default csd, enter widget edit mode.
Create 1 widget  \(I created a scroll number.\) duplicate it 4 times, creating 5 widgets in a right-slanted descending cluster.
Save the file as test.csd.
Select the 5 widgets \(I drag-selected.\)
Right-click the selection, select Center horizontally. 
I now see five widgets arranged horizontally, overlapping quite a bit. 
Type cmd-Z. Nothing happens on my system.
Select EDIT/UNDO. Nothing happens on my system.
Select qutecsound-f/quit
I get no option to save the file \(which of course was changed by the "Center Horizontally" command.\)
Double-click the csd icon to launch QTCS and open the file. I get the original configuration of diagonally descending widgets \(meaning that I was wrong: the widgets did not get written to the file after aligning them.\)
Select the widgets and right-click to "Center Vertically". Now the widgets are one on top of the other.
Type cmd-Z. Nothing happens.
Select EDIT/UNDO. The widgets are returned to their original position.
Select the widgets and execute Center Vertically again. The widgets are now one above the other.
Select FILE/RELOAD. nothing happens.
Quit \(no offer to save the file\)
Double-click the file icon. I get the widgets in the configuration before I centered vertically, so they did not get saved \(again I was wrong about this.\)
Select the lowest widget, move it down, away for the other four.
Select the five widgets, Execute align vertically. the widgets are one above the other, the lowest one at some distance from the other four.
Type CMD-Z. The widgets return to the diagonal position BEFORE I moved one of the widgets away from the other four.
Move the lowest widget to the right side of the other four.
Execute Distribute Horizontally.
Type CMD-Z. The widgets return to the diagonal position BEFORE I moved one of the widgets away from the other four.
